src/wine-host/editor.cpp

```diff
@@ -116,6 +123,31 @@ constexpr uint32_t xembed_focus_first = 1;
  */
 static const HCURSOR arrow_cursor = LoadCursor(nullptr, IDC_ARROW);
 
+/**
+ * List of keycodes that shall be forwarded to host (REAPER) when
+ * 'forward_key_events_to_host_' is true.
+ *
+ * NOTE: List needs to be in sorted order to work with binary search.
+ *       'spacebar' and 'function keys' were choosen, because those are what
+ *       is typical for REAPER to receive even when editor (wine) window has
+ *       focus (i.e. native behavior on all OS).
+ */
+const std::vector<int> keycodes_to_forward {
```

Comment:
Part of the reason why I didn't really want to look into myself (and I'd much rather have REAPER implement similar behavior as Bitwig) is that doing something like this is quite brittle, and doing it the right way is a bit involved. These key codes are specific to QWERTY keyboard layouts (and potentially other system settings). The proper way to do this is to map a list of `KEYSYM`s  to `KEYCODE`s first. See the keyboard section of the X11 spec here: https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html#Keyboards

(this should also be an `std::unordered_set` for constant time lookups, but I can fix those things up later)

Reply:
> Part of the reason why I didn't really want to look into myself (and I'd much rather have REAPER implement similar behavior as Bitwig) is that doing something like this is quite brittle, and doing it the right way is a bit involved. These key codes are specific to QWERTY keyboard layouts (and potentially other system settings). The proper way to do this is to map a list of `KEYSYM`s to `KEYCODE`s first. See the keyboard section of the X11 spec here: https://www.x.org/releases/X11R7.7/doc/xproto/x11protocol.html#Keyboards

I've just read through that section you linked (couple times actually). I guess I can see for why this would be the "proper way" of making this robust for any `KEYSYM`.

Maybe I just wrongly assumed that at least keys like `spacebar>` and `function keys` would be quite universally always the same `KEYCODE` on X11. At least that's the impression I got after finding things like this one:
https://gist.github.com/rickyzhang82/8581a762c9f9fc6ddb8390872552c250

I did find this library mentioned during my research: https://xkbcommon.org/
Using the `KEYCODE` directly just seemed the only feasible thing for me at the time without adding another dependency like this one. But I might have been mistaken of what can be done "natively" with `xcb`.


> (this should also be an `std::unordered_set` for constant time lookups, but I can fix those things up later)

Just had to read up on it. Yeah, I guess that would be the correct choice for a lookup scenario in general. I didn't even think of this, because I figured for such a small list that won't grow any further (assuming this keeps being just about `spacebar` and `function keys`), any kind of performance gains from a more sophisticated data structure than a simple array seemed negligible. Or if it's about avoiding any kind of issues due to "variable execution time" down the line, well, I'd assume even a simple for-loop for such a small list would be hard to notice.
But of course, I didn't make the measurement. Really just tried to learn something from this.